generateBins.py

```diff
@@ -25,69 +13,50 @@
 
-
 # TODO Determine the bins computationally
 # Following bins are for Rwanda
 POPULATION_BINS = [2125, 5640, 8989, 12108, 15577, 20289, 27629, 49378, 95262, 286928]
 
-# getting data ready for binning
-# data should be 1-dimensional array, python list or iterable
-myArray  = np.loadtxt("GIS\\rwa_population.asc", skiprows = 6)
-print (type(myArray))
-print(myArray.ndim)
-print(myArray.shape)
-print(myArray.size)
-array_1d = myArray.flatten()
-print (array_1d)
-print(array_1d.ndim)
-print(array_1d.shape)
-print(array_1d.size)
-#gvf = 0.0
-#nclasses = 20
-
-def goodness_of_variance_fit(array, classes):
-    # get the break points
-    classes = jenkspy.jenks_breaks(array, classes)
-
-    # do the actual classification
-    classified = np.array([classify(i, classes) for i in array])
-
-    # max value of zones
-    maxz = max(classified)
-
-    # nested list of zone indices
-    zone_indices = [[idx for idx, val in enumerate(classified) if zone + 1 == val] for zone in range(maxz)]
-
-    # sum of squared deviations from array mean
-    sdam = np.sum((array - array.mean()) ** 2)
-
-    # sorted polygon stats
-    array_sort = [np.array([array[index] for index in zone]) for zone in zone_indices]
-
-    # sum of squared deviations of class means
-    sdcm = sum([np.sum((classified - classified.mean()) ** 2) for classified in array_sort])
-
-    # goodness of variance fit
-    gvf = (sdam - sdcm) / sdam
-
-    return gvf
 
-
-def process(configuration, gisPath = ""):
+def process(configuration, gisPath=""):
     # Load the configuration
     cfg = load_configuration(configuration)
+    filename = os.path.join(gisPath, PFPR_FILE)
 
     # TODO Add the stuff for the population bins!
+    # GVF Implementation
+    binning = bin_asc(filename)
+
+    list_bins = []
+    list_GVF = []
+    # Assuming the range for bins to be between 5 and 30 (computing the optimal number of bins w.r.t. GVF), to avoid overfitting or underfitting of data
+    for x in range(5, 30):
+        GVF = goodness_of_variance_fit(binning, x)
+        # GVF less than 0.7 might not be the best fit for data; GVF ~ 0 (No fit), GVF ~ 1 (Perfect fit)
+        if GVF < 0.7:
+            print ("Bins=", x,"GVF = ", GVF, "Warning: GVF too low")
+        # Storing the values of bins and GVF in a list
+        list_bins.append(x)
+        list_GVF.append(GVF)
+    # computing difference between GVF values and storing them as (key, value) pairs in dictionary
+    diff_GVF = {list_bins[i]: abs(list_GVF[i] - list_GVF[i - 1]) for i in range(len(list_bins))}
+    for key, value in diff_GVF.items():
+        # taking such value of bin for which the difference between consecutive GVF values is greater than or equal to 0.1
+        if value >= 0.1:
+            print("Total number of bins", key)
+            # corresponding to the computed bin(key), finding the value of GVF
+            GVF = goodness_of_variance_fit(binning, key)
+            print("The value of Goodness of variance fit", GVF)
```

Comment:
Not quite sure where you are going here though - remember that once we know the number of bins with an acceptable GVF score (i.e., > 0.7 and with the heuristic being satisfied) then we want to use the bin values returned by Jenks in lieu of `POPULATION_BINS`.